Test failure: `ThriftOverHttpClientTest.testHelloServiceSync` and `testMessageLogsForException`

```
> Task :thrift0.13:shadedTest FAILED

ThriftOverHttpClientTest > testHelloServiceSync(ClientOptions, SerializationFormat, SessionProtocol) > com.linecorp.armeria.client.thrift.ThriftOverHttpClientTest.testHelloServiceSync(ClientOptions, SerializationFormat, SessionProtocol)[15] FAILED
    java.lang.IllegalArgumentException: expectedSize cannot be negative but was: -1
        at com.linecorp.armeria.internal.shaded.guava.collect.CollectPreconditions.checkNonnegative(CollectPreconditions.java:39)
        at com.linecorp.armeria.internal.shaded.guava.collect.ImmutableList.builderWithExpectedSize(ImmutableList.java:731)
        at com.linecorp.armeria.common.stream.DefaultStreamMessage.drainAll(DefaultStreamMessage.java:233)
        at com.linecorp.armeria.common.stream.DefaultStreamMessage.collectAll(DefaultStreamMessage.java:213)
        at com.linecorp.armeria.common.stream.DefaultStreamMessage.lambda$collect$1(DefaultStreamMessage.java:192)
        at com.linecorp.armeria.common.RequestContext.lambda$makeContextAware$3(RequestContext.java:546)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:384)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)

ThriftOverHttpClientTest > testMessageLogsForException(ClientOptions, SerializationFormat, SessionProtocol) > com.linecorp.armeria.client.thrift.ThriftOverHttpClientTest.testMessageLogsForException(ClientOptions, SerializationFormat, SessionProtocol)[40] FAILED
    java.lang.AssertionError: 
    Expecting:
      <java.lang.IllegalArgumentException: expectedSize cannot be negative but was: -1>
    to be an instance of:
      <org.apache.thrift.TApplicationException>
    but was:
      <"java.lang.IllegalArgumentException: expectedSize cannot be negative but was: -1
    	at com.linecorp.armeria.internal.shaded.guava.collect.CollectPreconditions.checkNonnegative(CollectPreconditions.java:39)
    	at com.linecorp.armeria.internal.shaded.guava.collect.ImmutableList.builderWithExpectedSize(ImmutableList.java:731)
    	at com.linecorp.armeria.common.stream.DefaultStreamMessage.drainAll(DefaultStreamMessage.java:233)
    	at com.linecorp.armeria.common.stream.DefaultStreamMessage.collectAll(DefaultStreamMessage.java:213)
    	at com.linecorp.armeria.common.stream.DefaultStreamMessage.lambda$collect$1(DefaultStreamMessage.java:192)
    	at com.linecorp.armeria.common.RequestContext.lambda$makeContextAware$3(RequestContext.java:546)
    	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:164)
    	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
    	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:384)
    	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
    	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
    	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
    	at java.lang.Thread.run(Thread.java:748)
    ">
        at com.linecorp.armeria.client.thrift.ThriftOverHttpClientTest.testMessageLogsForException(ThriftOverHttpClientTest.java:684)
```

--- a/core/src/main/java/com/linecorp/armeria/common/stream/AbstractStreamMessage.java
+++ b/core/src/main/java/com/linecorp/armeria/common/stream/AbstractStreamMessage.java
@@ -63,7 +63,7 @@ abstract class AbstractStreamMessage<T> implements StreamMessage<T> {
         requireNonNull(executor, "executor");
         requireNonNull(options, "options");
 
-        final SubscriptionImpl subscription = new SubscriptionImpl(this, subscriber, executor, options, null);
+        final SubscriptionImpl subscription = new SubscriptionImpl(this, subscriber, executor, options);
         final SubscriptionImpl actualSubscription = subscribe(subscription);
         if (actualSubscription != subscription) {
             // Failed to subscribe.
@@ -155,22 +155,17 @@ abstract class AbstractStreamMessage<T> implements StreamMessage<T> {
         private final boolean withPooledObjects;
         private final boolean notifyCancellation;
 
-        @Nullable
-        private final CompletableFuture<?> collectingFuture;
-
         private volatile boolean cancelRequested;
 
         @SuppressWarnings("unchecked")
         SubscriptionImpl(AbstractStreamMessage<?> publisher, Subscriber<?> subscriber,
-                         EventExecutor executor, SubscriptionOption[] options,
-                         @Nullable CompletableFuture<?> collectingFuture) {
+                         EventExecutor executor, SubscriptionOption[] options) {
             this.publisher = publisher;
             this.subscriber = (Subscriber<Object>) subscriber;
             this.executor = executor;
             this.options = options;
             withPooledObjects = containsWithPooledObjects(options);
             notifyCancellation = containsNotifyCancellation(options);
-            this.collectingFuture = collectingFuture;
         }
 
         Subscriber<Object> subscriber() {
@@ -199,10 +194,6 @@ abstract class AbstractStreamMessage<T> implements StreamMessage<T> {
             return withPooledObjects;
         }
 
-        boolean notifyCancellation() {
-            return notifyCancellation;
-        }
-
         boolean cancelRequested() {
             return cancelRequested;
         }
@@ -229,15 +220,6 @@ abstract class AbstractStreamMessage<T> implements StreamMessage<T> {
         // majority of cases.
         boolean needsDirectInvocation() {
             return executor.inEventLoop();
-        }
-
-        @Nullable
-        CompletableFuture<?> collectingFuture() {
-            return collectingFuture;
-        }
-
-        boolean isCollecting() {
-            return collectingFuture != null;
         }
 
         @Override
@@ -284,10 +266,6 @@ abstract class AbstractStreamMessage<T> implements StreamMessage<T> {
                 try {
                     if (subscription.notifyCancellation || !(cause instanceof CancelledSubscriptionException)) {
                         subscriber.onError(cause);
-                    }
-                    final CompletableFuture<?> collectingFuture = subscription.collectingFuture();
-                    if (collectingFuture != null) {
-                        collectingFuture.completeExceptionally(cause);
                     }
                     completionFuture.completeExceptionally(cause);
                 } catch (Throwable t) {

--- a/core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java
+++ b/core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java
@@ -184,16 +184,7 @@ public class DefaultStreamMessage<T> extends AbstractStreamMessageAndWriter<T> {
         if (subscriptionUpdater.compareAndSet(this, null, subscription)) {
             // We don't need to invoke onSubscribe() for NoopSubscriber.
             invokedOnSubscribe = true;
-            if (setState(State.CLOSED, State.CLEANUP)) {
-                final boolean withPooledObjects = subscription.withPooledObjects();
-                if (executor.inEventLoop()) {
-                    collectAll(collectingFuture, executor, withPooledObjects, true);
-                } else {
-                    executor.execute(() -> collectAll(collectingFuture, executor, withPooledObjects, false));
-                }
-            } else {
-                notifySubscriber();
-            }
+            notifySubscriber();
         } else {
             final Subscriber<Object> subscriber = this.subscription.subscriber();
             final Throwable cause;
@@ -205,77 +196,6 @@ public class DefaultStreamMessage<T> extends AbstractStreamMessageAndWriter<T> {
             collectingFuture.completeExceptionally(cause);
         }
         return collectingFuture;
-    }
-
-    private void collectAll(CompletableFuture<List<T>> collectingFuture, EventExecutor executor,
-                            boolean withPooledObjects, boolean directExecution) {
-        try {
-            collectingFuture.complete(drainAll(withPooledObjects));
-            if (directExecution) {
-                // The collectingFuture is not returned yet. We can guarantee that whenComplete() will be
-                // completed after executing the callbacks of collect() by rescheduling it.
-                executor.execute(() -> whenComplete().complete(null));
-            } else {
-                // We don't know whether the collectingFuture is returned or not at the moment. Just complete
-                // whenComplete() immediately.
-                whenComplete().complete(null);
-            }
-        } catch (Throwable throwable) {
-            collectingFuture.completeExceptionally(throwable);
-            whenComplete().completeExceptionally(throwable);
-        }
-    }
-
-    private List<T> drainAll(boolean withPooledObjects) throws Throwable {
-        // ClosedEvent was added to the queue
-        final int estimatedSize = queue.size() - 1;
-
-        final ImmutableList.Builder<T> builder = ImmutableList.builderWithExpectedSize(estimatedSize);
-        CloseEvent closeEvent = null;
-        Throwable closeCause = null;
-        for (;;) {
-            final Object o = queue.poll();
-            if (o == null) {
-                break;
-            }
-
-            if (o instanceof CloseEvent) {
-                closeEvent = (CloseEvent) o;
-                final Throwable cause = closeEvent.cause;
-                if (cause != null && closeCause == null) {
-                    closeCause = cause;
-                }
-                continue;
-            }
-
-            if (o instanceof AwaitDemandFuture) {
-                final AwaitDemandFuture future = (AwaitDemandFuture) o;
-                if (closeEvent == null) {
-                    future.complete(null);
-                } else {
-                    if (closeCause == null) {
-                        closeCause = ClosedStreamException.get();
-                    }
-                    future.completeExceptionally(closeCause);
-                }
-                continue;
-            }
-
-            @SuppressWarnings("unchecked")
-            T t = (T) o;
-            t = prepareObjectForNotification(t, withPooledObjects);
-            builder.add(t);
-        }
-
-        final List<T> objs = builder.build();
-        if (closeEvent == null || closeEvent == SUCCESSFUL_CLOSE) {
-            return objs;
-        } else {
-            for (T obj: objs) {
-                StreamMessageUtil.closeOrAbort(obj, closeCause);
-            }
-            throw closeCause;
-        }
     }
 
     /**
@@ -397,7 +317,11 @@ public class DefaultStreamMessage<T> extends AbstractStreamMessageAndWriter<T> {
 
     private void notifySubscriberOfCloseEvent0(SubscriptionImpl subscription, CloseEvent event) {
         try {
-            event.notifySubscriber(subscription, whenComplete());
+            if (subscription.isCollecting()) {
+                finishCollecting(event.cause, subscription);
+            } else {
+                event.notifySubscriber(subscription, whenComplete());
+            }
         } finally {
             subscription.clearSubscriber();
             Throwable cause = event.cause;
@@ -550,6 +474,24 @@ public class DefaultStreamMessage<T> extends AbstractStreamMessageAndWriter<T> {
         return true;
     }
 
+    private void finishCollecting(@Nullable Throwable cause, SubscriptionImpl subscription) {
+        @SuppressWarnings("unchecked")
+        final CompletableFuture<List<T>> collectingFuture =
+                (CompletableFuture<List<T>>) subscription.collectingFuture();
+        if (cause == null) {
+            if (collector == null) {
+                // No data was produced.
+                collectingFuture.complete(ImmutableList.of());
+            } else {
+                collectingFuture.complete(Collections.unmodifiableList(collector));
+            }
+            whenComplete().complete(null);
+        } else {
+            collectingFuture.completeExceptionally(cause);
+            whenComplete().completeExceptionally(cause);
+        }
+    }
+
     private void collectElements(SubscriptionImpl subscription) {
         if (collector == null) {
             collector = new ArrayList<>(INITIAL_CAPACITY);
@@ -575,9 +517,7 @@ public class DefaultStreamMessage<T> extends AbstractStreamMessageAndWriter<T> {
     @Override
     public final void close() {
         if (setState(State.OPEN, State.CLOSED)) {
-            if (!tryCollect(null)) {
-                addObjectOrEvent(SUCCESSFUL_CLOSE);
-            }
+            addObjectOrEvent(SUCCESSFUL_CLOSE);
         }
     }
 
@@ -599,53 +539,10 @@ public class DefaultStreamMessage<T> extends AbstractStreamMessageAndWriter<T> {
      */
     protected final boolean tryClose(Throwable cause) {
         if (setState(State.OPEN, State.CLOSED)) {
-            if (!tryCollect(cause)) {
-                addObjectOrEvent(new CloseEvent(cause));
-            }
+            addObjectOrEvent(new CloseEvent(cause));
             return true;
         }
         return false;
-    }
-
-    private boolean tryCollect(@Nullable Throwable cause) {
-        final SubscriptionImpl subscription = this.subscription;
-        if (subscription != null) {
-            @SuppressWarnings("unchecked")
-            final CompletableFuture<List<T>> collectingFuture =
-                    (CompletableFuture<List<T>>) subscription.collectingFuture();
-            if (collectingFuture != null) {
-                if (subscription.needsDirectInvocation()) {
-                    tryCollect(cause, collectingFuture);
-                } else {
-                    subscription.executor().execute(() -> {
-                        tryCollect(cause, collectingFuture);
-                    });
-                }
-                return true;
-            }
-        }
-        return false;
-    }
-
-    private void tryCollect(@Nullable Throwable cause, CompletableFuture<List<T>> collectingFuture) {
-        if (cause == null) {
-            // We don't need to invoke onSubscribe() for collect().
-            invokedOnSubscribe = true;
-            notifySubscriber0();
-            if (collector == null) {
-                // No data was produced.
-                collectingFuture.complete(ImmutableList.of());
-            } else {
-                collectingFuture.complete(Collections.unmodifiableList(collector));
-            }
-            whenComplete().complete(null);
-        } else {
-            if (setState(State.CLOSED, State.CLEANUP)) {
-                cleanupObjects();
-            }
-            collectingFuture.completeExceptionally(cause);
-            whenComplete().completeExceptionally(cause);
-        }
     }
 
     private boolean setState(State oldState, State newState) {

--- a/core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java
+++ b/core/src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java
@@ -20,9 +20,6 @@ import static com.linecorp.armeria.common.util.Exceptions.throwIfFatal;
 import static com.linecorp.armeria.internal.common.stream.InternalStreamMessageUtil.EMPTY_OPTIONS;
 import static java.util.Objects.requireNonNull;
 
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.List;
 import java.util.Queue;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
@@ -33,8 +30,6 @@ import org.jctools.queues.MpscChunkedArrayQueue;
 import org.reactivestreams.Subscriber;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-
-import com.google.common.collect.ImmutableList;
 
 import com.linecorp.armeria.common.annotation.UnstableApi;
 
@@ -88,8 +83,6 @@ public class DefaultStreamMessage<T> extends AbstractStreamMessageAndWriter<T> {
     private static final int INITIAL_CAPACITY = 32;
 
     private final Queue<Object> queue;
-    @Nullable
-    private List<T> collector;
 
     @Nullable
     private Throwable cleanupCause;
@@ -173,31 +166,6 @@ public class DefaultStreamMessage<T> extends AbstractStreamMessageAndWriter<T> {
      */
     protected void subscribe0(EventExecutor executor, SubscriptionOption[] options) {}
 
-    @Override
-    public CompletableFuture<List<T>> collect(EventExecutor executor, SubscriptionOption... options) {
-        requireNonNull(executor, "executor");
-        requireNonNull(options, "options");
-
-        final CompletableFuture<List<T>> collectingFuture = new CompletableFuture<>();
-        final SubscriptionImpl subscription =
-                new SubscriptionImpl(this, NoopSubscriber.get(), executor, options, collectingFuture);
-        if (subscriptionUpdater.compareAndSet(this, null, subscription)) {
-            // We don't need to invoke onSubscribe() for NoopSubscriber.
-            invokedOnSubscribe = true;
-            notifySubscriber();
-        } else {
-            final Subscriber<Object> subscriber = this.subscription.subscriber();
-            final Throwable cause;
-            if (subscriber instanceof AbortingSubscriber) {
-                cause = ((AbortingSubscriber<Object>) subscriber).cause();
-            } else {
-                cause = new IllegalStateException("subscribed by other subscriber already");
-            }
-            collectingFuture.completeExceptionally(cause);
-        }
-        return collectingFuture;
-    }
-
     /**
      * Invoked whenever a new demand is requested.
      */
@@ -218,7 +186,7 @@ public class DefaultStreamMessage<T> extends AbstractStreamMessageAndWriter<T> {
         SubscriptionImpl subscription = this.subscription;
         if (subscription == null) {
             final SubscriptionImpl newSubscription = new SubscriptionImpl(
-                    this, AbortingSubscriber.get(cause), ImmediateEventExecutor.INSTANCE, EMPTY_OPTIONS, null);
+                    this, AbortingSubscriber.get(cause), ImmediateEventExecutor.INSTANCE, EMPTY_OPTIONS);
             if (subscriptionUpdater.compareAndSet(this, null, newSubscription)) {
                 // We don't need to invoke onSubscribe() for AbortingSubscriber because it's just a placeholder.
                 invokedOnSubscribe = true;
@@ -317,11 +285,7 @@ public class DefaultStreamMessage<T> extends AbstractStreamMessageAndWriter<T> {
 
     private void notifySubscriberOfCloseEvent0(SubscriptionImpl subscription, CloseEvent event) {
         try {
-            if (subscription.isCollecting()) {
-                finishCollecting(event.cause, subscription);
-            } else {
-                event.notifySubscriber(subscription, whenComplete());
-            }
+            event.notifySubscriber(subscription, whenComplete());
         } finally {
             subscription.clearSubscriber();
             Throwable cause = event.cause;
@@ -436,12 +400,6 @@ public class DefaultStreamMessage<T> extends AbstractStreamMessageAndWriter<T> {
     }
 
     private boolean notifySubscriberWithElements(SubscriptionImpl subscription) {
-        if (subscription.isCollecting()) {
-            // collect() was called. Should drain queued elements in order to invoke onRemoval() immediately.
-            collectElements(subscription);
-            return true;
-        }
-
         final Subscriber<Object> subscriber = subscription.subscriber();
         if (demand == 0) {
             return false;
@@ -472,34 +430,6 @@ public class DefaultStreamMessage<T> extends AbstractStreamMessageAndWriter<T> {
             inOnNext = false;
         }
         return true;
-    }
-
-    private void finishCollecting(@Nullable Throwable cause, SubscriptionImpl subscription) {
-        @SuppressWarnings("unchecked")
-        final CompletableFuture<List<T>> collectingFuture =
-                (CompletableFuture<List<T>>) subscription.collectingFuture();
-        if (cause == null) {
-            if (collector == null) {
-                // No data was produced.
-                collectingFuture.complete(ImmutableList.of());
-            } else {
-                collectingFuture.complete(Collections.unmodifiableList(collector));
-            }
-            whenComplete().complete(null);
-        } else {
-            collectingFuture.completeExceptionally(cause);
-            whenComplete().completeExceptionally(cause);
-        }
-    }
-
-    private void collectElements(SubscriptionImpl subscription) {
-        if (collector == null) {
-            collector = new ArrayList<>(INITIAL_CAPACITY);
-        }
-
-        @SuppressWarnings("unchecked")
-        final T o = (T) queue.remove();
-        collector.add(prepareObjectForNotification(o, subscription.withPooledObjects()));
     }
 
     private void notifyAwaitDemandFuture() {

--- a/core/src/main/java/com/linecorp/armeria/common/stream/DeferredStreamMessage.java
+++ b/core/src/main/java/com/linecorp/armeria/common/stream/DeferredStreamMessage.java
@@ -314,7 +314,7 @@ public class DeferredStreamMessage<T> extends AbstractStreamMessage<T> {
         }
 
         final SubscriptionImpl newSubscription = new SubscriptionImpl(
-                this, AbortingSubscriber.get(cause), ImmediateEventExecutor.INSTANCE, EMPTY_OPTIONS, null);
+                this, AbortingSubscriber.get(cause), ImmediateEventExecutor.INSTANCE, EMPTY_OPTIONS);
         downstreamSubscriptionUpdater.compareAndSet(this, null, newSubscription);
 
         final StreamMessage<T> upstream = this.upstream;
@@ -368,7 +368,7 @@ public class DeferredStreamMessage<T> extends AbstractStreamMessage<T> {
         final DefaultStreamMessage<?> streamMessage = new DefaultStreamMessage<>();
         streamMessage.close();
         return new SubscriptionImpl(streamMessage, NoopSubscriber.get(), ImmediateEventExecutor.INSTANCE,
-                                    EMPTY_OPTIONS, null);
+                                    EMPTY_OPTIONS);
     }
 
     private final class ForwardingSubscriber implements Subscriber<T> {

